### ISSUE TITLE
Detect plug-in/-out of monitor

### ISSUE THREAD

**Issue by genofire** [emersion/mako#281]
**Version:** 1.4.1
**DesktopEnviroment:** sway (wayland only)

Mako does not show any notify on any output if monitor disappear, even if it appears again.

Here a little testscript
```bash
swaymsg output eDP-1 disable
sleep 5
swaymsg output eDP-1 enable
notify-send "after a monitor exists again"
makoctl list
```

in the list the notify exists ....

----
another issue / nice to have would be to manage output by makoctl ...

**Comment by vilhalmer** [comment]
What version of mako?

**Comment by Inc0ns1st3nt** [comment]
I think I am experiencing  a similar issue, where I use an external monitor, and keep my laptop close.
For that purpose, I also had my `lid close behaviour` set to `do nothing`.
When I close my laptop, then connect to the monitor I think this issue crop up.
My lid close behaviour is set to do nothing.

**Comment by PopeRigby** [comment]
I'm having the same issue.

OS: Arch Linux
Mako version: 1.4.1-2

**Comment by donnex** [comment]
I think I have the same issue as well. It happens when I leave my machine idle and the monitor goes to sleep. When I come back mako won't show any notifications until I kill it and start it again. I just use a single display.

OS: Arch Linux
Mako version: 1.4.1-2

I don't know if this helps at all but here is the `strace` output when I try to send a notification when this happens.

```
recvmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\1\0\1\0\0\0\0\2\0\0\0\250\0\0\0\1\1o\0\36\0\0\0", iov_len=24}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 24
recvmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="/org/freedesktop/Notifications\0\0"..., iov_len=160}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 160
sendmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\2\1\1\227\10\0\0\274\1\0\0\37\0\0\0\5\1u\0\2\0\0\0\6\1s\0\7\0\0\0"..., iov_len=48}, {iov_base="\222\10\0\0<!DOCTYPE node PUBLIC \"-//fr"..., iov_len=2199}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 2247
recvmsg(3, {msg_namelen=0}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=5, events=POLLIN}], 4, -1) = 1 ([{fd=3, revents=POLLIN}])
recvmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\1\0\1T\0\0\0\3\0\0\0\260\0\0\0\1\1o\0\36\0\0\0", iov_len=24}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 24
recvmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="/org/freedesktop/Notifications\0\0"..., iov_len=252}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 252
timerfd_settime(6, TFD_TIMER_ABSTIME, {it_interval={tv_sec=0, tv_nsec=0}, it_value={tv_sec=51765, tv_nsec=475055940}}, NULL) = 0
sendmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\2\1\1\4\0\0\0\275\1\0\0\37\0\0\0\5\1u\0\3\0\0\0\6\1s\0\7\0\0\0"..., iov_len=48}, {iov_base="\200\0\0\0", iov_len=4}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 52
recvmsg(3, {msg_namelen=0}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=5, events=POLLIN}], 4, -1) = 1 ([{fd=6, revents=POLLIN}])
read(6, "\1\0\0\0\0\0\0\0", 8)          = 8
timerfd_settime(6, TFD_TIMER_ABSTIME, {it_interval={tv_sec=0, tv_nsec=0}, it_value={tv_sec=55262, tv_nsec=889605208}}, NULL) = 0
sendmsg(3, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\4\1\1\10\0\0\0\276\1\0\0x\0\0\0\1\1o\0\36\0\0\0/org/fre"..., iov_len=136}, {iov_base="\200\0\0\0\1\0\0\0", iov_len=8}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 144
poll([{fd=3, events=POLLIN}, {fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=5, events=POLLIN}], 4, -1
```
This is quite problematic sine you don't really know if and when it happen and important notifications can easily be missed.

Is there any way to debug this?

**Comment by vilhalmer** [comment]
This issue is the top of my fix wishlist, I just haven't had a chance to give reproducing it another go. I've been unable to reproduce it when actually in a position to debug it, though I think I did manage to trigger it once while rearranging displays.

I suspect that mako is still processing notifications totally normally from a d-bus perspective when it occurs, since the sending tool doesn't hang like tends to happen when notifications aren't being pulled off the bus in a timely fashion. It seems to either just never realize that an output of the same name as a previously-existing one has reappeared, or is somehow keeping around an output object with no actual association to a wl_output. The latter is more far-fetched since I would expect it to crash. Neither of them explains why it doesn't reproduce consistently for all users (i.e. me).

Does this happen every time your display goes to sleep, or just occasionally?

**Comment by PopeRigby** [comment]
It happens every single time for me.

**Comment by donnex** [comment]
For me this issue started a couple of weeks ago when I switched out my monitor setup. Went from a dual display setup with one HDMI monitor and one with DisplayPort. Now I only use a single DisplayPort monitor. I never had this issue before.

It seems to happen for me pretty much every time I leave the computer running and the monitor has gone to sleep. `swayidle` runs `output * dpms off`.

 Is there anything we can do to help with debugging to figure out what is happening?

**Comment by vilhalmer** [comment]
It's pretty much impossible to introspect mako's display management state except via attaching a debugger. If you want to make sure your build has debug symbols and try to catch it in action, by all means!

**Comment by alonso-m** [comment]
I've been experiencing this for few months. If there is any info that I can provide to help reproducing let me know. I can reproduce consistently.

**Comment by PopeRigby** [comment]
How do I make sure mako has debug symbols/build it with them?

**Comment by PopeRigby** [comment]
@vilhalmer How do I make sure mako has debug symbols/build it with them?

**Comment by vilhalmer** [comment]
If you build it manually using the instructions in the readme you should end up with debug symbols, then you'll just need to run that binary instead of the one shipped by your distro while you wait for it to happen again. You can verify that the symbols are present by running the built binary under gdb and checking if you get function names when you run `bt full`. When mako gets into the bad state you should be able to generate a core file by sending it SIGQUIT, or `gcore` which comes with gdb. Good luck!

**Comment by donnex** [comment]
@vilhalmer Thanks for the detailed information!

I haven't done any work with gdb before so this was a first for me. I think I was able to get a core file when mako was stuck in a bad state.

Cloned the repo and checked out tag v1.4.1 and built it. Then started it manually.

I was able to send notifications without any problems. To trigger this bad state when mako was running but no notifications shows up I only used a sleep of a few seconds and then send a notification. Before the notification was sent I unplugged my only monitor (displayport). I think the same thing happens when my monitor goes to sleep and sometimes when I unsuspend the computer.

The core file can be downloaded at https://donnex.net/core.505262

**Comment by vilhalmer** [comment]
Thanks! Hopefully this will shed some light. In the meantime if you don't mind, could you try reproducing on master to see if the bug is still present? We've rearranged a bunch of the code for multiple surface support so it's possible we fixed it by accident. :crossed_fingers:

**Comment by donnex** [comment]
Great .. let's hope the core file contains what's needed :)

I just tried with a fresh build of master but I could not reproduce the issue. At least not the same way. To be really sure I rebuilt 1.4.1 and got into a broken state the first try.

So it seems that something in master has fixed the issue :) At least this one I can reproduce.

I'll keep master running for a while to see if it it happens again for some other reason.
